depends/CMakeLists.txt

```diff
@@ -17,3 +17,62 @@ option(CLSOCKET_SHARED "Build clsocket lib as shared." OFF)
 option(CLSOCKET_DEP_ONLY "Build for use inside other CMake projects as dependency." ON)
 add_subdirectory(clsocket)
 ide_folder(clsocket "Depends")
+
+# assemble environment args to pass on to dependency projects
+get_cmake_property(vars CACHE_VARIABLES)
+foreach(var ${vars})
+  if(var MATCHES "^CMAKE_"
+     AND NOT var MATCHES "^CMAKE_CACHE"
+     AND NOT var MATCHES "^CMAKE_HOME"
+     AND NOT var MATCHES "^CMAKE_PROJECT")
+    list(APPEND CL_ARGS "-D${var}=${${var}}")
+  endif()
+  if(var MATCHES "^ZLIB")
+    list(APPEND CL_ARGS "-D${var}=${${var}}")
+  endif()
+endforeach()
+
+include(ExternalProject)
+
+if(WIN32)
+  set(EXPAT_LIB_NAME_SUFFIX "MD")
+  set(XLSXIO_C_FLAGS "${CMAKE_C_FLAGS} /DXML_STATIC")
+else()
+  set(EXPAT_LIB_NAME_SUFFIX "")
+  set(XLSXIO_C_FLAGS "${CMAKE_C_FLAGS} -DXML_STATIC")
+endif()
+
+set(LIBEXPAT_INSTALL_DIR ${CMAKE_CURRENT_BINARY_DIR}/libexpat)
```

Comment:
I tried building without setting up the new submodules and got a somewhat-unhelpful error message (where the suggested solutions aren't applicable to users who have forgotten this step):

```
CMake Error at /usr/share/cmake-3.16/Modules/ExternalProject.cmake:2630 (message):
  No download info given for 'libexpat_project' and its source directory:

   /path/to/dfhack/depends/libexpat/expat

  is not an existing non-empty directory.  Please specify one of:

   * SOURCE_DIR with an existing non-empty directory
   * DOWNLOAD_COMMAND
   * URL
   * GIT_REPOSITORY
   * SVN_REPOSITORY
   * HG_REPOSITORY
   * CVS_REPOSITORY and CVS_MODULE
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/ExternalProject.cmake:3236 (_ep_add_download_command)
  depends/CMakeLists.txt:48 (ExternalProject_Add)
```

There's an existing check for some other submodule files here: https://github.com/DFHack/dfhack/blob/f2b0f012c9895aa3fd743ad50fab9e6fb78b479d/CMakeLists.txt#L181

Do you have thoughts on a clean way to perform a similar check here? I was considering a check for the CMakeLists.txt in each folder before the call to ExternalProject_Add, but was wondering if you had better ideas.

Another option, of course, is to leave this as-is and document the error message in Compile.rst. My concern here is mainly that this is the first time we've added submodules in a while, and the first time using ExternalProject with them, so people with existing DFHack clones might not realize how to fix this issue if they run into it.

Reply:
Adding additional checks to the top-level CMakeLists.txt seems like a good idea -- I'd rather add it there than in depends/CMakeLists.txt since we already have the error message in place directing people to Compile.rst, which is still applicable here.